kubetest/kubeadmdind/kubeadm_dind.go

```diff
@@ -0,0 +1,527 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// This file implements a Kubernetes kubetest deployer based on the scripts
+// in the Mirantis/kubeadm-dind-cluster github repo. These scripts are used
+// to create a multinode, containerized Kubernetes cluster that is running
+// inside a Prow DinD container (i.e. Docker-in-Docker-in-Docker, or DinDinD
+// configuration).
+
+package kubeadmdind
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"time"
+
+	"k8s.io/test-infra/kubetest/process"
+)
+
+var (
+	// Names that are fixed in the Kubeadm DinD scripts
+	kubeMasterPrefix = "kube-master"
+	kubeNodePrefix   = "kube-node"
+
+	// Systemd service logs to collect on the host container
+	hostServices = []string{
+		"docker",
+	}
+
+	// Docker commands to run on the host container and embedded node
+	// containers for log dump
+	dockerCommands = []struct {
+		cmd     string
+		logFile string
+	}{
+		{"docker images", "docker_images.log"},
+		{"docker ps -a", "docker_ps.log"},
+	}
+
+	// Systemd service logs to collect on the master and worker embedded
+	// node containers for log dump
+	systemdServices = []string{
+		"kubelet",
+		"docker",
+	}
+	masterKubePods = []string{
+		"kube-apiserver",
+		"kube-scheduler",
+		"kube-controller-manager",
+		"kube-proxy",
+		"etcd",
+	}
+	nodeKubePods = []string{
+		"kube-proxy",
+		"kube-dns",
+	}
+
+	// Where to look for (nested) container log files on the node containers
+	nodeLogDir = "/var/log"
+
+	// Path to Kubernetes source tree
+	kubernetesDir = "/workspace/k8s.io/kubernetes"
+
+	// kubeadm-dind-cluster script
+	kdcScript = "/workspace/github.com/leblancd/kubeadm-dind-cluster/fixed/dind-cluster-stable.sh"
+
+	// Maximum number of nodes to create
+	maxNumNodes = 10
```

Comment:
why?

Reply:
This 'maxNumNodes' can be eliminated if I remove the 'numNodes' kubetest option and use a fixed number of worker nodes for this test. (This was a sanity limit check on the numNodes config setting... it could have been less restrictive).